src/coreclr/vm/siginfo.cpp

```diff
@@ -3895,7 +3927,14 @@ MetaSig::CompareElementType(
             IfFailThrow(CorSigUncompressElementType_EndPtr(pSig1, pEndSig1, &callingConvention1));
             CorElementType callingConvention2 = ELEMENT_TYPE_MAX; // initialize to illegal
             IfFailThrow(CorSigUncompressElementType_EndPtr(pSig2, pEndSig2, &callingConvention2));
-            if (callingConvention1 != callingConvention2)
+
+            // Calling conventions are generally treated as custom modifiers.
+            // When callers request calling conventions to be ignored, we also ignore
+            // calling conventions and this is okay. It is okay because calling conventions,
+            // when more than one is defined (for example, SuppressGCTransition), all
+            // become encoded as custom modifiers.
+            if (!state->IgnoreCustomModifiers
+                && callingConvention1 != callingConvention2)
```

Comment:
```suggestion
            if (state->IgnoreCustomModifiers ? 
                (NormalizeFnPtrCallingConvention(callingConvention1) != NormalizeFnPtrCallingConvention(callingConvention2))
                (callingConvention1 != callingConvention2))
```
We only want to ignore the calling convention bits that map to modopts, the rest should still match. It matches how we treat the calling convention bits in the rest of the system.

```
static uint32_t NormalizeFnPtrCallingConvention(uint32_t callConv)
{
    LIMITED_METHOD_CONTRACT;
    
    // Only have an unmanaged\managed status, and not the unmanaged CALLCONV_ value.
    switch (uCallConv & IMAGE_CEE_CS_CALLCONV_MASK)
    {
        case IMAGE_CEE_CS_CALLCONV_C:
        case IMAGE_CEE_CS_CALLCONV_STDCALL:
        case IMAGE_CEE_CS_CALLCONV_THISCALL:
        case IMAGE_CEE_CS_CALLCONV_FASTCALL:
            // Strip the calling convention.
            uCallConv &= ~IMAGE_CEE_CS_CALLCONV_MASK;
            // Normalize to unmanaged.
            uCallConv |= IMAGE_CEE_CS_CALLCONV_UNMANAGED;
    }

    return uCallConv;
}
```

I took this logic from https://github.com/dotnet/runtime/blob/main/src/coreclr/vm/siginfo.cpp#L1769-L1780. It should be replaced with a call to `NormalizeFnPtrCallingConvention` as well.

Reply:
The test for this is: `delegate* unmanaged<void>` and ` delegate* <void>`. They should not match with IgnoreCustomModifiers, but I think they do match today.